src/eko/io/struct.py

```diff
@@ -587,126 +170,58 @@ def permissions(self):
-
-        self._operators[mu2] = op
-        return op
-
-    def __setitem__(self, mu2: float, op: Operator, compress: bool = True):
-        """Set operator for given :math:`Q^2`.
-
-        The operator is automatically dumped on disk.
-
-        Parameters
-        ----------
-        q2 : float
-            :math:`Q^2` value labeling the operator to be set
-        op : Operator
-            the retrieved operator
-        compress : bool
-            whether to save the operator compressed or not (default: `True`)
-
-        """
-        self.access.assert_writeable()
-
-        if not isinstance(op, Operator):
-            raise ValueError("Only an Operator can be added to an EKO")
-
-        without_err = op.error is None
-        oppath = self.paths.opnewpath(mu2, compress=compress, without_err=without_err)
 
-        with open(oppath, "wb") as fd:
-            without_err2 = op.save(fd, compress)
-        assert without_err == without_err2
+    def load_recipes(self, recipes: List[Recipe]):
+        """Load recipes in bulk."""
+        for recipe in recipes:
+            # leverage auto-save
+            if isinstance(recipe, Evolution):
+                self.recipes[recipe] = None
+            else:
+                self.recipes_matching[recipe] = None
 
-        self._operators[mu2] = op
-
-    def __delitem__(self, q2: float):
-        """Drop operator from memory.
-
-        This method only drops the operator from memory, and it's not expected
-        to do anything else.
-
-        Autosave is done on set, and explicit saves are performed by the
-        computation functions.
-
-        If a further explicit save is required, repeat explicit assignment::
-
-            eko[q2] = eko[q2]
-
-        This is only useful if the operator has been mutated in place, that in
-        general should be avoided, since the operator should only be the result
-        of a full computation or a library manipulation.
-
-
-        Parameters
-        ----------
-        q2 : float
-            the value of :math:`Q^2` for which the corresponding operator
-            should be dropped
-
-        """
-        self._operators[q2] = None
+    # operator management
+    # -------------------
+    def __getitem__(self, ep: EPoint) -> Optional[Operator]:
+        r"""Retrieve operator for given evolution point."""
+        return self.operators[Target.from_ep(ep)]
+
+    def __setitem__(self, ep: EPoint, op: Operator):
+        """Set operator associated to an evolution point."""
+        self.operators[Target.from_ep(ep)] = op
+
+    def __delitem__(self, ep: EPoint):
+        """Drop operator from memory."""
+        del self.operators[Target.from_ep(ep)]
+
+    def __delattr__(self, name: str):
+        """Empty an inventory cache."""
+        attr = getattr(self, name)
+        if isinstance(attr, Inventory):
+            attr.empty()
+        else:
+            super().__delattr__(name)
```

Comment:
@AleCandido what is the use case of this line? why would I delete an eko attribute? and can I even do it on an ordinary Python object? I think this should be replaced by a `ValueError` or similar ...

Reply:
Yes, you can do it on an ordinary Python object. In particular, on whatever Python object.
After all, Python is a dynamic language, otherwise it would be C.

Indeed, it is just the counterpart of setting new attributes: as you can set an attribute to an instance, you can also delete it.
And setting new attributes only to `self` and only inside `__init__` is just a convention and nothing else. `self` is not even a keyword, and you could choose whatever name you prefer inside `__init__`. And `__init__` would be a perfectly normal function, apart that is called right after the creation of a new instance, with the arguments provided to the constructor.

Reply:
So, the `else` branch is there just to avoid modification of the default behavior of Python objects, apart from the specific case in which I want to use `del` with a different meaning.

In any case, this syntax:
```py
del eko.operators
```
it is only syntactic sugar for:
```py
eko.operators.empty()
```

I would not worry too much about what `del` does in general (you're just surprised because you're encountering a new Python operator for the first time).

P.S.: btw `del x.y` is `__delattr__` and `del x[y]` is `__delitem__`, and they are quite different, despite using the same keyword: the first is defined for all objects, while the second is only used by some containers (`list`, `dict`, and other mutable containers)

Reply:
fine